File: research/label_optimizer/cost_model.py
"""
Square Root Cost Model (v4.6)

Nonlinear market impact model for realistic cost estimation.
Impact = η × σ × √(Q / V)
"""

from dataclasses import dataclass
import numpy as np


@dataclass
class SquareRootCostModel:
    """
    Square Root Market Impact Model.

    Attributes:
        base_fee: Exchange fee (0.1% = 0.001)
        base_slippage: Minimum slippage (0.01% = 0.0001)
        impact_coeff: Impact coefficient (η)
        slippage_cap: Maximum slippage cap (0.5% = 0.005)
        min_profit_buffer: Required profit multiplier over cost
    """

    base_fee: float = 0.001  # 0.10%
    base_slippage: float = 0.0001  # 0.01%
    impact_coeff: float = 0.1
    slippage_cap: float = 0.005  # 0.50%
    min_profit_buffer: float = 1.5

    def get_slippage(
        self,
        volatility: float,
        trade_size: float = 0.01,
        avg_volume: float = 1.0,
    ) -> float:
        """
        Calculate slippage using Square Root Law.

        slippage = base + η × σ × √(trade_size / volume)

        Args:
            volatility: Current volatility (σ)
            trade_size: Order size as fraction of volume
            avg_volume: Average volume for normalization

        Returns:
            Total slippage (capped)
        """
        if avg_volume <= 0:
            return self.slippage_cap

        volume_ratio = trade_size / avg_volume
        if volume_ratio < 0:
            volume_ratio = 0
        impact = self.impact_coeff * volatility * np.sqrt(volume_ratio)

        total_slippage = self.base_slippage + impact
        return min(total_slippage, self.slippage_cap)

    def get_total_cost(
        self,
        volatility: float,
        trade_size: float = 0.01,
        avg_volume: float = 1.0,
    ) -> float:
        """
        Calculate round-trip total cost (entry + exit).

        Returns:
            Total cost as fraction of position
        """
        slippage = self.get_slippage(volatility, trade_size, avg_volume)
        return 2 * (self.base_fee + slippage)

    def is_viable(
        self,
        pt_mult: float,
        sigma: float,
        volume_ratio: float = 1.0,
    ) -> bool:
        """
        Fee Trap Guard: Check if trade is profitable.

        Expected Profit = PT × σ
        Required = Total Cost × Buffer

        Args:
            pt_mult: Profit target multiplier
            sigma: Current volatility
            volume_ratio: Trade size / average volume

        Returns:
            True if expected profit exceeds required threshold
        """
        expected_profit = pt_mult * sigma
        total_cost = self.get_total_cost(sigma, trade_size=volume_ratio)
        required = total_cost * self.min_profit_buffer

        return expected_profit > required

    def get_breakeven_pt(self, sigma: float, volume_ratio: float = 1.0) -> float:
        """
        Calculate minimum PT multiplier for breakeven.

        Args:
            sigma: Current volatility
            volume_ratio: Trade size / average volume

        Returns:
            Minimum PT multiplier needed
        """
        if sigma <= 0:
            return float("inf")

        total_cost = self.get_total_cost(sigma, trade_size=volume_ratio)
        return (total_cost * self.min_profit_buffer) / sigma

File: etl/internal/backtest/cost_model.go
package backtest

import "math"

// CostModel implements the Square Root Market Impact Model.
//
// Impact = η × σ × √(Q / V)
//
// Where:
//   - η: Impact coefficient
//   - σ: Volatility
//   - Q: Order size
//   - V: Average volume
type CostModel struct {
	BaseFee      float64 // Base exchange fee (e.g., 0.001 = 0.1%)
	BaseSlippage float64 // Minimum slippage (e.g., 0.0001 = 0.01%)
	ImpactCoeff  float64 // Market impact coefficient (η)
	SlippageCap  float64 // Maximum slippage cap
}

// NewCostModel creates a new cost model from config.
func NewCostModel(cfg Config) *CostModel {
	return &CostModel{
		BaseFee:      cfg.BaseFee,
		BaseSlippage: cfg.BaseSlippage,
		ImpactCoeff:  cfg.ImpactCoeff,
		SlippageCap:  cfg.SlippageCap,
	}
}

// DefaultCostModel returns a cost model with sensible defaults.
func DefaultCostModel() *CostModel {
	return &CostModel{
		BaseFee:      0.001,  // 0.10% exchange fee
		BaseSlippage: 0.0001, // 0.01% minimum slippage
		ImpactCoeff:  0.1,    // Impact coefficient
		SlippageCap:  0.005,  // 0.50% maximum slippage
	}
}

// GetSlippage calculates slippage using the Square Root Law.
//
// slippage = base_slippage + η × σ × √(trade_size / avg_volume)
//
// The result is capped at SlippageCap.
func (c *CostModel) GetSlippage(volatility, tradeSize, avgVolume float64) float64 {
	if avgVolume <= 0 {
		return c.SlippageCap
	}

	volumeRatio := tradeSize / avgVolume
	if volumeRatio < 0 {
		volumeRatio = 0
	}

	impact := c.ImpactCoeff * volatility * math.Sqrt(volumeRatio)
	totalSlippage := c.BaseSlippage + impact

	// Cap the slippage
	if totalSlippage > c.SlippageCap {
		return c.SlippageCap
	}

	return totalSlippage
}

// GetTotalCost calculates round-trip cost (entry + exit).
//
// Total Cost = 2 × (base_fee + slippage)
func (c *CostModel) GetTotalCost(volatility, tradeSize, avgVolume float64) float64 {
	slippage := c.GetSlippage(volatility, tradeSize, avgVolume)
	return 2 * (c.BaseFee + slippage)
}

// GetEntryCost calculates the cost for entry only.
func (c *CostModel) GetEntryCost(volatility, tradeSize, avgVolume float64) float64 {
	slippage := c.GetSlippage(volatility, tradeSize, avgVolume)
	return c.BaseFee + slippage
}

// ApplyEntrySlippage adjusts the entry price for slippage.
//
// Long: entry_price × (1 + slippage) - pay more
// Short: entry_price × (1 - slippage) - receive less
func (c *CostModel) ApplyEntrySlippage(price float64, direction Direction, volatility, tradeSize, avgVolume float64) float64 {
	slippage := c.GetSlippage(volatility, tradeSize, avgVolume)

	if direction == DirectionLong {
		return price * (1 + slippage)
	}
	// Short
	return price * (1 - slippage)
}

// ApplyExitSlippage adjusts the exit price for slippage.
//
// Long: exit_price × (1 - slippage) - receive less
// Short: exit_price × (1 + slippage) - pay more to cover
func (c *CostModel) ApplyExitSlippage(price float64, direction Direction, volatility, tradeSize, avgVolume float64) float64 {
	slippage := c.GetSlippage(volatility, tradeSize, avgVolume)

	if direction == DirectionLong {
		return price * (1 - slippage)
	}
	// Short
	return price * (1 + slippage)
}

// CalculatePnL computes the P&L for a trade.
//
// For Long: (exit - entry) / entry
// For Short: (entry - exit) / entry
//
// Returns both gross (before fees) and net (after fees) P&L.
func (c *CostModel) CalculatePnL(entryPrice, exitPrice float64, direction Direction) (grossPnL, netPnL float64) {
	if direction == DirectionLong {
		grossPnL = (exitPrice - entryPrice) / entryPrice
	} else {
		grossPnL = (entryPrice - exitPrice) / entryPrice
	}

	// Subtract fees (applied to both entry and exit)
	netPnL = grossPnL - 2*c.BaseFee

	return grossPnL, netPnL
}

// IsViable checks if a trade is likely profitable after costs.
//
// Expected Profit = PT × σ
// Required = Total Cost × Buffer
//
// Returns true if expected profit > required.
func (c *CostModel) IsViable(ptMult, volatility, tradeSize, avgVolume, buffer float64) bool {
	expectedProfit := ptMult * volatility
	totalCost := c.GetTotalCost(volatility, tradeSize, avgVolume)
	required := totalCost * buffer

	return expectedProfit > required
}
